### PR TITLE
Compute tag_name a maximum of once per document owner

### LINKED ISSUE #1471: Consider precomputing Element#tagName.
According to jgraham:

It looks like someone didn't finish the namespace implementation for elements
At the moment they store something called tag_name
But they should store local_name and tag_name should be computed

### PR DESCRIPTION
**Pull request by mitchhentges**
Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data:
- [x] `./mach build -d` does not report any errors
- [X] `./mach test-tidy --faster` does not report any errors
- [X] These changes fix #1471 (github issue number if applicable).

Either:
- [X] These changes do not require tests because no new functionality was added, just a reorganization and caching of existing functionality

Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11318)

### CODE REVIEW

**Comment by nox** [comment]
-S-awaiting-review +S-needs-code-changes 

Nice work, @mitchhentges.

<details><summary><i>Previously, highfive wrote…</i></summary>
> Heads up! This PR modifies the following files:
>  * @kichjang: components/script/dom/element.rs
</details>


---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[components/script/dom/element.rs, line 68 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11318#-KIIv6S0B-BWfk7SZN2Q:-KIIv6S1IB5c8f5iLgPY:-1688006804) ([raw file](https://github.com/servo/servo/blob/d63d07a5b56cc95b54f13f9b58f15c8a84edf472/components/script/dom/element.rs#L68)):*
> ```Rust
> use dom::validation::Validatable;
> use dom::virtualmethods::{VirtualMethods, vtable_for};
> use heapsize::{HeapSizeOf};
> ```

Nit: remove braces.

---

*[components/script/dom/element.rs, line 1305 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11318#-KIIzPAd7UAL-4pPy2IY:-KIIzPAd7UAL-4pPy2IZ:-2097957028) ([raw file](https://github.com/servo/servo/blob/d63d07a5b56cc95b54f13f9b58f15c8a84edf472/components/script/dom/element.rs#L1305)):*
> ```Rust
>                 qualified_name.to_ascii_uppercase()
>             } else {
>                 qualified_name.into_owned()
> ```

There is no need for `into_owned` here now that there is an implementation of `From<Cow<str>>` for `Atom`.

---

*[components/script/dom/element.rs, line 2104 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11318#-KIIzbWZ4fwf_5FEY2KK:-KIIzbWZ4fwf_5FEY2KL:697964147) ([raw file](https://github.com/servo/servo/blob/d63d07a5b56cc95b54f13f9b58f15c8a84edf472/components/script/dom/element.rs#L2104)):*
> ```Rust
>     fn adopting_steps(&self, old_doc: &Document) {
>         self.super_type().unwrap().adopting_steps(old_doc);
>         self.tag_name.clear();
> ```

Just noticed that you can limit the `clear` call to cases such:

```rust
document_from_node(self).is_html_document() != old_doc.is_html_document()
```

---

*[components/script/dom/element.rs, line 2552 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11318#-KIJ-9m1q_rZNdAZpuAu:-KIJ-9m1q_rZNdAZpuAv:-885977558) ([raw file](https://github.com/servo/servo/blob/d63d07a5b56cc95b54f13f9b58f15c8a84edf472/components/script/dom/element.rs#L2552)):*
> ```Rust
> // A holder for an element's "tag name", which will be lazily
> // resolved and cached. Should be reset when the document
> // owner change
> ```

Nit: three slashes and a final period.

---

*[components/script/dom/element.rs, line 2566 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11318#-KIJ-Ef6iJZmVrupKusw:-KIJ-Ef6iJZmVrupKusx:-441849261) ([raw file](https://github.com/servo/servo/blob/d63d07a5b56cc95b54f13f9b58f15c8a84edf472/components/script/dom/element.rs#L2566)):*
> ```Rust
> 
> impl TagName {
>     pub fn new() -> TagName {
> ```

Make it private.

---

*[components/script/dom/element.rs, line 2573 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11318#-KIJ-DQxSWxyPMCtcnh1:-KIJ-DQxSWxyPMCtcnh2:-441849261) ([raw file](https://github.com/servo/servo/blob/d63d07a5b56cc95b54f13f9b58f15c8a84edf472/components/script/dom/element.rs#L2573)):*
> ```Rust
>     /// initialized with the result of `cb` first.
>     #[allow(unsafe_code)]
>     pub fn or_init<F>(&self, cb: F) -> Atom
> ```

Make it private.

---

*[components/script/dom/element.rs, line 2587 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11318#-KIJ-IHK3KfIZLl8A4EW:-KIJ-IHK3KfIZLl8A4EX:-293687213) ([raw file](https://github.com/servo/servo/blob/d63d07a5b56cc95b54f13f9b58f15c8a84edf472/components/script/dom/element.rs#L2587)):*
> ```Rust
> 
>     // Clear the cached tag name, so that it will be re-calculated the
>     // next time that `or_init()` is called
> ```

Nit: period.

---

*[components/script/dom/element.rs, line 2589 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11318#-KIJ-Go-TIypGu9Cl0Ve:-KIJ-Go-TIypGu9Cl0Vf:-441849261) ([raw file](https://github.com/servo/servo/blob/d63d07a5b56cc95b54f13f9b58f15c8a84edf472/components/script/dom/element.rs#L2589)):*
> ```Rust
>     // next time that `or_init()` is called
>     #[allow(unsafe_code)]
>     pub fn clear(&self) {
> ```

Make it private.

---

*[components/script/dom/element.rs, line 2592 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11318#-KIJ-NoutOZv6yEhIZBC:-KIJ-NoutOZv6yEhIZBD:-2068107672) ([raw file](https://github.com/servo/servo/blob/d63d07a5b56cc95b54f13f9b58f15c8a84edf472/components/script/dom/element.rs#L2592)):*
> ```Rust
>         unsafe {
>             let ptr = self.ptr.get();
>             *ptr = None;
> ```

Nit: no need for the intermediate `ptr` variable.

---

*[components/script/dom/element.rs, line 2595 \[r1\]](https://reviewable.io:443/reviews/servo/servo/11318#-KIJ-JxFnLQBbLaIKZCo:-KIJ-JxFnLQBbLaIKZCp:713643203) ([raw file](https://github.com/servo/servo/blob/d63d07a5b56cc95b54f13f9b58f15c8a84edf472/components/script/dom/element.rs#L2595)):*
> ```Rust
>         }
>     }
> }
> ```

Nit: final empty line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11318#-:-KIJ-UZOmMwjB24WuzEh:1369464240)*

**Comment by nox** [comment]
@bors-servo try

-S-awaiting-review

@Ms2ger Do you know if there are tests for changing a element's owner document and checking its tag name?

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11318#-:-KIJ7FMQBnJHmQ3ieiYR:-999738065)*

**Comment by KiChjang** [comment]
Unit tests are failing:
```
failures:
---- size_of::size_element stdout ----
	thread 'size_of::size_element' panicked at 'Your changes have increased the stack size of commonly used DOM struct Element from 312 to 328. These structs are present in large quantities in the DOM, and increasing the size may dramatically affect our memory footprint. Please consider choosing a design which avoids this increase. If you feel that the increase is necessary, update to the new size in tests/unit/script/size_of.rs.', /home/travis/build/servo/servo/tests/unit/script/size_of.rs:42
stack backtrace:
   1:     0x7f0faf558190 - std::sys::backtrace::tracing::imp::write::h9fb600083204ae7f
   2:     0x7f0faf55babb - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::hca543c34f11229ac
   3:     0x7f0faf55b6ac - std::panicking::default_hook::hc2c969e7453d080c
   4:     0x7f0faf54f138 - std::panicking::rust_panic_with_hook::hfe203e3083c2b544
   5:     0x7f0faf55bd01 - std::panicking::begin_panic::h4889569716505182
   6:     0x7f0faf5500ba - std::panicking::begin_panic_fmt::h484cd47786497f03
   7:     0x7f0faed3e3c1 - script_tests::size_of::size_element::he2aeb340c66061cc
                        at /home/travis/build/servo/servo/components/servo/<std macros>:8
   8:     0x7f0faf533906 - _<F as std..boxed..FnBox<A>>::call_box::hebdedc6b66e381cc
   9:     0x7f0faf536037 - std::panicking::try::call::hdb2c8d43be906104
  10:     0x7f0faf56593b - __rust_try
  11:     0x7f0faf5658de - __rust_maybe_catch_panic
  12:     0x7f0faf53645e - _<F as std..boxed..FnBox<A>>::call_box::h214ad3593915cb36
  13:     0x7f0faf55a0d4 - std::sys::thread::Thread::new::thread_start::h6f266e069bf4ec2b
  14:     0x7f0facdaf181 - start_thread
  15:     0x7f0fac8c647c - __clone
  16:                0x0 - <unknown>
---- size_of::size_div stdout ----
	thread 'size_of::size_div' panicked at 'Your changes have increased the stack size of commonly used DOM struct HTMLDivElement from 328 to 344. These structs are present in large quantities in the DOM, and increasing the size may dramatically affect our memory footprint. Please consider choosing a design which avoids this increase. If you feel that the increase is necessary, update to the new size in tests/unit/script/size_of.rs.', /home/travis/build/servo/servo/tests/unit/script/size_of.rs:44
stack backtrace:
   1:     0x7f0faf558190 - std::sys::backtrace::tracing::imp::write::h9fb600083204ae7f
   2:     0x7f0faf55babb - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::hca543c34f11229ac
   3:     0x7f0faf55b6ac - std::panicking::default_hook::hc2c969e7453d080c
   4:     0x7f0faf54f138 - std::panicking::rust_panic_with_hook::hfe203e3083c2b544
   5:     0x7f0faf55bd01 - std::panicking::begin_panic::h4889569716505182
   6:     0x7f0faf5500ba - std::panicking::begin_panic_fmt::h484cd47786497f03
   7:     0x7f0faed3ea41 - script_tests::size_of::size_div::hca661af76121f93c
                        at /home/travis/build/servo/servo/components/servo/<std macros>:8
   8:     0x7f0faf533906 - _<F as std..boxed..FnBox<A>>::call_box::hebdedc6b66e381cc
   9:     0x7f0faf536037 - std::panicking::try::call::hdb2c8d43be906104
  10:     0x7f0faf56593b - __rust_try
  11:     0x7f0faf5658de - __rust_maybe_catch_panic
  12:     0x7f0faf53645e - _<F as std..boxed..FnBox<A>>::call_box::h214ad3593915cb36
  13:     0x7f0faf55a0d4 - std::sys::thread::Thread::new::thread_start::h6f266e069bf4ec2b
  14:     0x7f0facdaf181 - start_thread
  15:     0x7f0fac8c647c - __clone
  16:                0x0 - <unknown>
---- size_of::size_span stdout ----
	thread 'size_of::size_span' panicked at 'Your changes have increased the stack size of commonly used DOM struct HTMLSpanElement from 328 to 344. These structs are present in large quantities in the DOM, and increasing the size may dramatically affect our memory footprint. Please consider choosing a design which avoids this increase. If you feel that the increase is necessary, update to the new size in tests/unit/script/size_of.rs.', /home/travis/build/servo/servo/tests/unit/script/size_of.rs:45
stack backtrace:
   1:     0x7f0faf558190 - std::sys::backtrace::tracing::imp::write::h9fb600083204ae7f
   2:     0x7f0faf55babb - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::hca543c34f11229ac
   3:     0x7f0faf55b6ac - std::panicking::default_hook::hc2c969e7453d080c
   4:     0x7f0faf54f138 - std::panicking::rust_panic_with_hook::hfe203e3083c2b544
   5:     0x7f0faf55bd01 - std::panicking::begin_panic::h4889569716505182
   6:     0x7f0faf5500ba - std::panicking::begin_panic_fmt::h484cd47786497f03
   7:     0x7f0faed3ed81 - script_tests::size_of::size_span::h01b04b425602691c
                        at /home/travis/build/servo/servo/components/servo/<std macros>:8
   8:     0x7f0faf533906 - _<F as std..boxed..FnBox<A>>::call_box::hebdedc6b66e381cc
   9:     0x7f0faf536037 - std::panicking::try::call::hdb2c8d43be906104
  10:     0x7f0faf56593b - __rust_try
  11:     0x7f0faf5658de - __rust_maybe_catch_panic
  12:     0x7f0faf53645e - _<F as std..boxed..FnBox<A>>::call_box::h214ad3593915cb36
  13:     0x7f0faf55a0d4 - std::sys::thread::Thread::new::thread_start::h6f266e069bf4ec2b
  14:     0x7f0facdaf181 - start_thread
  15:     0x7f0fac8c647c - __clone
  16:                0x0 - <unknown>
---- size_of::size_htmlelement stdout ----
	thread 'size_of::size_htmlelement' panicked at 'Your changes have increased the stack size of commonly used DOM struct HTMLElement from 328 to 344. These structs are present in large quantities in the DOM, and increasing the size may dramatically affect our memory footprint. Please consider choosing a design which avoids this increase. If you feel that the increase is necessary, update to the new size in tests/unit/script/size_of.rs.', /home/travis/build/servo/servo/tests/unit/script/size_of.rs:43
stack backtrace:
   1:     0x7f0faf558190 - std::sys::backtrace::tracing::imp::write::h9fb600083204ae7f
   2:     0x7f0faf55babb - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::hca543c34f11229ac
   3:     0x7f0faf55b6ac - std::panicking::default_hook::hc2c969e7453d080c
   4:     0x7f0faf54f138 - std::panicking::rust_panic_with_hook::hfe203e3083c2b544
   5:     0x7f0faf55bd01 - std::panicking::begin_panic::h4889569716505182
   6:     0x7f0faf5500ba - std::panicking::begin_panic_fmt::h484cd47786497f03
   7:     0x7f0faed3e701 - script_tests::size_of::size_htmlelement::h3acca56fab8b1b38
                        at /home/travis/build/servo/servo/components/servo/<std macros>:8
   8:     0x7f0faf533906 - _<F as std..boxed..FnBox<A>>::call_box::hebdedc6b66e381cc
   9:     0x7f0faf536037 - std::panicking::try::call::hdb2c8d43be906104
  10:     0x7f0faf56593b - __rust_try
  11:     0x7f0faf5658de - __rust_maybe_catch_panic
  12:     0x7f0faf53645e - _<F as std..boxed..FnBox<A>>::call_box::h214ad3593915cb36
  13:     0x7f0faf55a0d4 - std::sys::thread::Thread::new::thread_start::h6f266e069bf4ec2b
  14:     0x7f0facdaf181 - start_thread
  15:     0x7f0fac8c647c - __clone
  16:                0x0 - <unknown>
```

**Comment by Ms2ger** [comment]
I'm sorry, but this is r- from me because of the unsafe code.

File: components/script/dom/element.rs [hunk 1]
```diff
@@ -2535,3 +2549,49 @@ impl AtomicElementFlags {
         self.0.fetch_or(flags.bits() as usize, Ordering::Relaxed);
     }
 }
+
+/// A holder for an element's "tag name", which will be lazily
+/// resolved and cached. Should be reset when the document
+/// owner changes.
+#[derive(JSTraceable)]
+struct TagName {
+    ptr: UnsafeCell<Option<Atom>>,
```
Comment (KiChjang) [inline components/script/dom/element.rs]: I agree with @Ms2ger, why is this an `UnsafeCell` rather than a `Cell` or `RefCell`?

**Comment by nox** [comment]
@Ms2ger It's exactly the same scheme as `UniqueId` in node.rs, I'm the one who mentioned UnsafeCell to not waste space for something that is just lazily-initiated. We could just use `RefCell`, but I would rather not make all elements 16 bytes larger.

@KiChjang We can't use `Cell`, `Atom` is not `Copy`.

**Comment by nox** [comment]
@Ms2ger Would you agree to use `mitochondria::CloningCell`? :)

https://github.com/nox/mitochondria/blob/master/src/cloning.rs

**Comment by Ms2ger** [comment]
I'd much rather not. Perhaps we should introduce a `NodeInfo` struct first, so the size increase would be much less important.

**Comment by mitchhentges** [comment]
Is this good to be worked on again? Is the consensus that using a `NodeInfo` is the direction that this should go?

**Comment by nox** [comment]
Seems like it.

**Comment by jdm** [comment]
The consensus on IRC when we discussed this was to use a DOMRefCell for this PR and do the NodeInfo work separately.

**Comment by mitchhentges** [comment]
OK, sounds good

**Comment by mitchhentges** [comment]
I'm again getting failing tests about increasing the memory footprint of common elements, but I'm assuming that that's expected

File: components/script/dom/element.rs [hunk 2]
```diff
@@ -2681,3 +2695,47 @@ impl AtomicElementFlags {
...
+/// resolved and cached. Should be reset when the document
+/// owner changes.
+#[derive(JSTraceable)]
+struct TagName {
+    ptr: DOMRefCell<Option<Atom>>,
+}
+
+impl HeapSizeOf for TagName {
+    #[allow(unsafe_code)]
+    fn heap_size_of_children(&self) -> usize {
```
Comment (Ms2ger) [inline components/script/dom/element.rs]: You should be able to just `derive` the `HeapSizeOf` trait.

**Comment by Ms2ger** [comment]
Yeah, just fix the expected numbers. Thanks!

**Comment by nox** [comment]
@bors-servo try

I think this needs expectation changes, so `try` instead of `r+`.

**Comment by mitchhentges** [comment]
Just [these four](https://github.com/servo/servo/pull/11318/files#diff-2eceaf470f0400ee8f81237aa62cf2c7R43) expectation changes, right?

**Comment by nox** [comment]
No, we didn't implement the right thing before, so there should be new unexpected passing tests. The fact that there are none means new tests need to be written.

**Comment by Ms2ger** [comment]
Hold on, did this change behaviour?

**Comment by nox** [comment]
Mmmh guess I was wrong about that, sorry it has been a long time since I looked at that PR. :/

File: components/script/dom/element.rs [hunk 3]
```diff
@@ -2681,3 +2694,40 @@ impl AtomicElementFlags {
...
+}
+
+impl TagName {
+    fn new() -> TagName {
+        TagName { ptr: DOMRefCell::new(None) }
+    }
+
+    /// Retrieve a copy of the current inner value. If it is `None`, it is
+    /// initialized with the result of `cb` first.
+    #[allow(unsafe_code)]
```
Comment (Ms2ger) [inline components/script/dom/element.rs]: Let's remove the two `#[allow(unsafe_code)]` attributes; should then be good to go.

**Comment by mitchhentges** [comment]
18th time the charm :wink:

**Comment by nox** [comment]
@bors-servo r+

Thanks again for your contribution!

**Pull request merged**